driver-core/src/main/com/mongodb/internal/session/BaseClientSessionImpl.java

```
@@ -33,16 +35,29 @@
     private final ClientSessionOptions options;
     private BsonDocument clusterTime;
     private BsonTimestamp operationTime;
+    private ServerAddress pinnedMongosAddress;
     private volatile boolean closed;
 
     public BaseClientSessionImpl(final ServerSessionPool serverSessionPool, final Object originator, final ClientSessionOptions options) {
         this.serverSessionPool = serverSessionPool;
         this.serverSession = serverSessionPool.get();
         this.originator = originator;
         this.options = options;
+        this.pinnedMongosAddress = null;
         closed = false;
     }
 
+    @Override
+    @Nullable
+    public ServerAddress getPinnedMongosAddress() {
+        return pinnedMongosAddress;
+    }
+
+    @Override
+    public void setPinnedMongosAddress(@Nullable final ServerAddress address) {
+        pinnedMongosAddress = address;
```

To be safe, this should throw an exception if pinnedMongoAddress is not currently null.

---

`pinnedMongosAddress` can be set to null when a transaction is started.

---

Yea, that's true.  So it would have to be something like:

   isTrue(address == null || pinnedMongosAddress == null)

---

Done